theano/gof/cc.py

```diff
@@ -503,12 +503,32 @@ def fetch_variables(self):
         self.inputs = fgraph.inputs
         self.outputs = fgraph.outputs
 
+        self.node_order = self.schedule(fgraph)
+
         # list(fgraph.variables)
-        # We need to include the not used inputs in our variables,
+        # We need to include the unused inputs in our variables,
         # otherwise we can't pass them to the module.
         self.variables = [var for var in self.inputs if not len(var.clients)]
         self.variables += graph.variables(self.inputs, self.outputs)
 
+        # This adds a hidden input which is the context for each node
+        # that needs it
+        self.contexts = dict()
+        for node in self.node_order:
+            ctx = node.run_context()
+            if ctx is not graph.NoContext:
+                # try to avoid creating more than one variable for the
+                # same context.
+                if ctx in self.contexts:
+                    var = self.contexts[ctx]
+                    assert var.type == node.context_type
+                    var.clients.append((node, 'context'))
```

Comment:
Since I add those clients at the very end of optimization (basically just before compiling the C code), No optimizer will ever see these clients, which is why it will be fine.

However I do have to add it to the clients list otherwise CLinker complains.

Also about it creating a new client, this is only for nodes that have a context defined.  Specifically CPU nodes, and old-backend nodes will not have this.